packages/flutter/test/widgets/nested_scroll_view_test.dart

```diff
@@ -676,6 +678,233 @@ void main() {
+                  controller: scrollController,
+                  key: globalKey,
+                  body: ListView.builder(
+                    itemBuilder: (BuildContext context, int index) {
+                      return Text('$index');
+                    },
+                  ),
+                  headerSliverBuilder: (_, __) => <Widget>[const SliverAppBar()],
+                ),
+              ),
+            ),
+          ));
+
+          // The scroll gesture should be taken where in the inner body, so the whole scroll view is scrolled.
+          final TestGesture gesture = await tester.startGesture(const Offset(0, kToolbarHeight + 1));
+          await gesture.moveBy(Offset(0, -scrollExtent));
+
+          await tester.pump();
+
+          expect(
+            globalKey.currentState.innerController.position.pixels +
+                globalKey.currentState.outerController.position.pixels,
+            scrollExtent,
+          );
+
+          // The outer scroll view cannot scroll past its height, which should be kToolbarHeight as a SliverAppBar represents the outer body.
+          expect(
+            globalKey.currentState.outerController.position.pixels,
+            min(kToolbarHeight, scrollExtent),
+          );
+
+          // The inner controller should only start scrolling once the outer controller has been scrolled to its full extent.
+          expect(
+            globalKey.currentState.innerController.position.pixels,
+            max(0, scrollExtent - kToolbarHeight),
+          );
+        });
+  }
+
+  testWidgets('NestedScrollViewState.outerController should correspond to NestedScrollView.controller', (WidgetTester tester) async {
+    final GlobalKey<NestedScrollViewState> globalKey = GlobalKey();
+    final ScrollController scrollController = ScrollController();
+
+    await tester.pumpWidget(Directionality(
+      textDirection: TextDirection.ltr,
+      child: Localizations(
+        locale: const Locale('en', 'US'),
+        delegates: const <LocalizationsDelegate<dynamic>>[
+          DefaultMaterialLocalizations.delegate,
+          DefaultWidgetsLocalizations.delegate,
+        ],
+        child: MediaQuery(
+          data: const MediaQueryData(),
+          child: NestedScrollView(
+            controller: scrollController,
+            key: globalKey,
+            body: ListView.builder(
+              itemBuilder: (BuildContext context, int index) {
+                return Text('$index');
+              },
+            ),
+            headerSliverBuilder: (_, __) => <Widget>[const SliverAppBar()],
+          ),
+        ),
+      ),
+    ));
+
+    // The scrollController should be equal to the outerController defined in the state.
+    expect(
+      scrollController.position,
+      globalKey.currentState.outerController.position,
+    );
+  });
+
+  // Test all cases in both directions for both controllers.
+  const List<List<dynamic>> nestedScrollViewStateControllersCases = <List<dynamic>>[
```

Comment:
This makes the test really complicated to follow. Maybe separate them out into individual test cases and factor common logic out into methods called from the test cases.